ddcommon-ffi/src/array_queue.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Error;
+use anyhow::Context;
+use std::{ffi::c_void, ptr::NonNull};
+
+#[derive(Debug)]
+#[repr(C)]
+// A simple wrapper around crossbeam_queue::ArrayQueue<*mut c_void>, which is a lock free
+// bounded multi-producer and multi-consumer (MPMC) queue.
+pub struct ArrayQueue {
+    // The actual type here should be NonNull<crossbeam_queue::ArrayQueue<*mut c_void>>.
+    // However, cbindgen does not use the module name crossbeam_queue to generate the C header, and
+    // above will be generated into ArrayQueue<*void>. So we use NonNull<ArrayQueue> here and cast
+    // it to the correct type in the FFI functions. Also, we use NonNull instead of Box to avoid
+    // getting into trouble with the drop implementation.
+    inner: NonNull<ArrayQueue>,
+    item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+}
+
+unsafe impl Sync for ArrayQueue {}
+unsafe impl Send for ArrayQueue {}
+
+impl ArrayQueue {
+    pub fn new(
+        capacity: usize,
+        item_delete_fn: Option<unsafe extern "C" fn(*mut c_void) -> c_void>,
+    ) -> anyhow::Result<Self, anyhow::Error> {
+        if capacity == 0 {
+            return Err(anyhow::anyhow!("capacity must be greater than 0"));
+        }
+        let item_delete_fn = item_delete_fn.context("item_delete_fn must be non-null")?;
+
+        let internal_queue: crossbeam_queue::ArrayQueue<*mut c_void> =
+            crossbeam_queue::ArrayQueue::new(capacity);
+        // # Safety: internal_queue must be non-null If the memory allocation had failed, the
+        // program would panic.
+        let inner = unsafe {
+            NonNull::new_unchecked(Box::into_raw(Box::new(internal_queue)) as *mut ArrayQueue)
+        };
```

Comment:
This happens so rarely, it would be cleaner to just 
```suggestion
        let inner = NonNull::new(Box::into_raw(Box::new(internal_queue)) as *mut ArrayQueue)?;
```